compiler/AST/type.cpp

```diff
@@ -591,6 +591,9 @@ void initPrimitiveTypes() {
   dtLocale                             = new AggregateType(AGGREGATE_RECORD);
   dtLocale->symbol                     = new TypeSymbol("locale", dtLocale);
 
+  dtLocaleClass                        = new AggregateType(AGGREGATE_RECORD);
```

Comment:
Why is this called `dtLocale*Class*` but then is pointed at an AGGREGATE_RECORD rather than AGGREGATE_CLASS type?

Reply:
Not-so-well-informed answer: The compiler does some special-casing for the lifetime and memory management for `LocaleModel`s. My guess is that making them `AGGREGATE_RECORD` helps with the autoDestroy process. Note that before this PR, `class _locale` is also an `AGGREGATE_RECORD`.

One way to test (I'll do in few mins) is to switch them to class and see if we leak any memory for example.

Reply:
Huh, OK...  I think that switch + test is probably worthwhile.  It'd be nice to lie to the compiler the least amount possible.

Thinking about it a bit more and looking through the code more, I also feel a little surprised that the compiler would need to know about the internal baselocale class type at all.  Specifically, I'd think that the `locale` record type should be able to implement itself however it needs to and the compiler would be none the wiser.  It looks like dtLocaleClass is barely used by the compiler, where in one case your comment suggests it's no-op-ish, and I'm wondering whether the other case might relate to it being an AGGREGATE_RECORD rather than an AGGREGATE_CLASS (?).  Maybe put a different way: if I were to implement a `locale` type that did something completely different within its implementation, why would the compiler care (or need to care) what fields it had or what they stored?  What goes wrong if the compiler doesn't know anything at all about dtLocaleClass?

Reply:
Didn't notice any change w.r.t cleanup. `dtLocale` was added in https://github.com/chapel-lang/chapel/pull/11405 as an `AGGREGATE_RECORD`. There may be still something that is about the weird status of `locale` before this PR which may have been resolved with this PR.

The question is whether I should pursue changing this to CLASS, or keep it for now not to keep the PR self-contained and try changing to CLASS afterwards.

Reply:
I think the reason dtLocale had AGGREGATE_RECORD before is that we were trying to fake to the compiler that locales had value semantics even though classes (used to implement them) didn't.  So it was a white lie to the compiler that was required to treat them as being value-like in most contexts.

I think the thing I'd try to do in this PR is remove the need to introduce dtLocaleClass into the compiler at all.  Having read through the rest of the code now, it doesn't look like the compiler should need to know about the type, and I don't think there's value in introducing it here just to pull it out later.  Specifically:

* most of the compiler code is involved in hooking it and setting it up
* the conditional in scopeResolve() is commented in a way indicating that it never triggers (which seems believable to me)
* `typeHasInfiniteBorrowLifetime` doesn't seem like it should be worrying about unmanaged classes, and I think all instances of the locale class instances are unmanaged?

Reply:
We don't seem to lose anything by removing it.

Reply:
That's about the most exciting news I've heard all day!  (it's been a slow day)